diff --git a/devices/xiaomi.vacuum/firmwarebuilder/flasher.py b/devices/xiaomi.vacuum/firmwarebuilder/flasher.py
index d3fa10e..97c95c5 100755
--- a/devices/xiaomi.vacuum/firmwarebuilder/flasher.py
+++ b/devices/xiaomi.vacuum/firmwarebuilder/flasher.py
@@ -27,6 +27,9 @@ import socketserver
 import threading
 from time import sleep
 import miio
+import socket
+import fcntl
+import struct
 
 
 class ThreadedHTTPServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
@@ -64,7 +67,8 @@ def md5(fname):
 
 
 def findIP():
-    return ((([ip for ip in socket.gethostbyname_ex(socket.gethostname())[2] if not ip.startswith('127.')] or [[(s.connect(('8.8.8.8', 53)), s.getsockname()[0], s.close()) for s in [socket.socket(socket.AF_INET, socket.SOCK_DGRAM)]][0][1]]) + ['no IP found'])[0])
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    return socket.inet_ntoa(fcntl.ioctl(s.fileno(), 0x8915, struct.pack('256s', bytes('eth1', 'utf-8')))[20:24])
 
 
 def discover_devices():
diff --git a/devices/xiaomi.vacuum/firmwarebuilder/imagebuilder.sh b/devices/xiaomi.vacuum/firmwarebuilder/imagebuilder.sh
index 48ef419..687dab4 100755
--- a/devices/xiaomi.vacuum/firmwarebuilder/imagebuilder.sh
+++ b/devices/xiaomi.vacuum/firmwarebuilder/imagebuilder.sh
@@ -367,7 +367,7 @@ if [ $UNPROVISIONED -eq 1 ]; then
 
         mkdir $IMG_DIR/opt/unprovisioned
         cp $BASEDIR/unprovisioned/start_wifi.sh $IMG_DIR/opt/unprovisioned
-        chmod +x ./opt/unprovisioned/start_wifi.sh
+        chmod +x $IMG_DIR/opt/unprovisioned/start_wifi.sh
 
         sed -i 's/exit 0//' $IMG_DIR/etc/rc.local
         cat $BASEDIR/unprovisioned/rc.local >> $IMG_DIR/etc/rc.local
